### DISCUSSION TITLE
part-ii/2016-24/

### DISCUSSION THREAD

**Discussion by tripos-comments** [tripos-education/maths-tripos-questions#129] [Question comments]
Part II 2016 Automata and Formal Languages: Paper 1, Section II, F 

https://questions.tripos.org/part-ii/2016-24/

**Comment by dropdbassett** [comment]
Question was mostly bookwork, but the final part (is it necessarily recursive?) was trickier than I was expecting. I would be interested if anyone had a simpler solution than what I came up with:
<details>
<summary>Is it necessarily recursive?</summary>

For input $n$, first calculate what $n$ represents as shortlex $\mathbb{N} \to \mathbb{N}^2$. If this is $(a,b)$, then define 

$$f(n) = \begin{cases}a &\text{ if the  } a\text{th register machine halts on }a\text{ in at most }b\text{ steps}\\k &\text{ otherwise},\end{cases}$$ 

where $k$ is some fixed element of $\mathbb{K}$. Since shortlex $\mathbb{N} \to \mathbb{N}^2$ is surjective, $\mathrm{Im} f = \mathbb{K}$.
</details>

**Reply by artie2000** [reply]
<details>
<summary>An alternative solution which I found intuitive:</summary>

The idea is to set $f|_\mathbb{K}=\text{Id}$ and $f|_{\mathbb{N}\setminus\mathbb{K}}=c_{n_0}$ constant. Then $\text{im} f=\mathbb{K}\cup\{n_0\}$, which it's easy to show is undecidable.

Formally, apply s-m-n on $f_{f(a),1}(b)=g(a,b)$, where

$$g(a,b)\coloneqq\begin{cases}f_{a,1}(b)&\text{if }a\text{ codes a program and }f_{a,1}(a)\downarrow\\\uparrow&\text{o/w}\end{cases}$$

</details>

**Reply by thenewtonium** [reply]
<details><summary>I don't think the "formally" bit works? (@FM22)</summary>

Since s-m-n only guarantees that $f_{f(a),1} = f_{a,1}$ for $a\in\mathbb{K}$, not that $f(a) = a$ here. And similarly, on $\mathbb{N}\setminus\mathbb{K}$, just because $f(a)$ codes an infinite loop for all $a$, it doesn't necessarily hold that that $f(a)$ codes the *same* infinite loop program for all $a\in\mathbb{N}\setminus\mathbb{K}$, surely?

**Reply by thenewtonium** [reply]
<details><summary>@dropdbassett maybe not ‘simpler’ but more intuitive? (Using Church's thesis):</summary>

You can get a computable function $f$ with image $\mathbb{K}$ by letting $f(n)$ be essentially "the $n$th the element of $\mathbb{K}$ when we enumerate $\mathbb{K}$". I.e. given input $n$ we use parallelism to compute the $f_{m,1}(m)$'s, and after $n+1$ halt of these parallel processes halt, we output the $m$ corresponding to the final one to halt.

EDIT: Come to think of it, this is basically the same idea as the proof of r.e. => image of a computable function.
</summary>

**Reply by artie2000** [reply]
> I don't think the "formally" bit works? (@FM22)

Oops, you are right. Good point!

**Reply by dropdbassett** [reply]
@thenewtonium That does seem better. Thanks!
